File: backend/app/schemas/pve.py
"""PVE 连接相关 schema。"""

import ipaddress
import re
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field, field_validator

_PVE_HOST_RE = re.compile(r"^[A-Za-z0-9](?:[A-Za-z0-9._-]*[A-Za-z0-9])?$")


def _validate_pve_host(value: str) -> str:
    """Validate a host/IP without allowing URL/path injection."""
    host = value.strip()
    if host.startswith("[") and host.endswith("]"):
        host = host[1:-1].strip()
    if not host:
        raise ValueError("host 不能为空")
    try:
        ipaddress.ip_address(host)
    except ValueError:
        if not _PVE_HOST_RE.fullmatch(host):
            raise ValueError("host 只允许 IP 或主机名,不能包含 URL、路径或特殊字符")
    return host


class PveConnectionCreate(BaseModel):
    name: str = Field(min_length=1, max_length=128)
    host: str = Field(min_length=1, max_length=255)
    port: int = Field(8006, ge=1, le=65535)
    token_id: str = Field(
        min_length=1, max_length=255, description="形如 root@pam!tokenid"
    )
    token_secret: str = Field(min_length=1, max_length=255)
    verify_ssl: int = Field(1, ge=0, le=1)
    enabled: int = Field(1, ge=0, le=1)
    description: str | None = Field(None, max_length=500)

    @field_validator("token_id")
    @classmethod
    def _check_token_id(cls, v: str) -> str:
        if "!" not in v or "@" not in v:
            raise ValueError("Token ID 格式应为 user@realm!tokenid,如 root@pam!dcn")
        return v

    @field_validator("host")
    @classmethod
    def _check_host(cls, v: str) -> str:
        return _validate_pve_host(v)


class PveConnectionUpdate(BaseModel):
    name: str | None = Field(None, min_length=1, max_length=128)
    host: str | None = Field(None, min_length=1, max_length=255)
    port: int | None = Field(None, ge=1, le=65535)
    token_id: str | None = Field(None, min_length=1, max_length=255)
    token_secret: str | None = Field(None, min_length=1, max_length=255)
    verify_ssl: int | None = Field(None, ge=0, le=1)
    enabled: int | None = Field(None, ge=0, le=1)
    description: str | None = Field(None, max_length=500)

    @field_validator("host")
    @classmethod
    def _check_host(cls, v: str | None) -> str | None:
        if v is None:
            return None
        return _validate_pve_host(v)

    @field_validator("token_id")
    @classmethod
    def _check_token_id(cls, v: str | None) -> str | None:
        if v is None:
            return None
        v = v.strip()
        if "!" not in v or "@" not in v:
            raise ValueError("Token ID 格式应为 user@realm!tokenid,如 root@pam!dcn")
        return v


class PveConnectionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    host: str
    port: int
    token_id: str
    verify_ssl: int
    enabled: int
    description: str | None = None


class PveGuestBindingUpdate(BaseModel):
    ip_address: str | None = Field(default=None, max_length=255)
    os_system: str | None = Field(default=None, pattern="^(linux|windows|unknown)$")
    ssh_port: int = Field(default=22, ge=1, le=65535)
    winrm_port: int = Field(default=5985, ge=1, le=65535)
    username: str = Field(min_length=1, max_length=255)
    password: str | None = Field(default=None, max_length=1000)
    ssh_key: str | None = Field(default=None, max_length=20000)
    enabled: bool = True


class PveGuestBindingResponse(BaseModel):
    id: int | None = None
    connection_id: int
    guest_type: str
    vmid: int
    ip_address: str | None = None
    os_system: str
    # True 表示 os_system=linux 是排除法兜底假设(未确证),前端应自动探测纠正
    os_assumed: bool = False
    # 持久化的精确 OS 名(QGA 不可用时的兜底显示值,见迁移 0038)
    os_name: str | None = None
    ssh_port: int
    winrm_port: int
    username: str | None = None
    has_password: bool = False
    has_ssh_key: bool = False
    enabled: bool
    configured: bool = False
    last_test_status: str | None = None
    last_test_error: str | None = None
    last_test_at: datetime | None = None
    created_at: datetime | None = None
    qga_enabled: bool = False
    qga_available: bool = False
    qga_ip_address: str | None = None
    qga_os_system: str | None = None
    qga_os_name: str | None = None
    qga_disks: list[dict] = Field(default_factory=list)
    winrm_available: bool = False

File: frontend/src/types/pve.ts
export interface PveConnection {
  id: number
  name: string
  host: string
  port: number
  token_id: string
  verify_ssl: number
  enabled: number
  description: string | null
  created_at: string | null
}

export interface PveNode {
  node: string
  status: string
  cpu?: number
  maxcpu?: number
  mem?: number
  maxmem?: number
  level?: string
  [key: string]: unknown
}

export interface PveFilesystemUsage {
  mount: string
  filesystem_type?: string | null
  total_bytes: number
  used_bytes?: number | null
}

export interface PveGuest {
  node: string
  type: 'qemu'
  vmid: number
  name: string
  status: string
  cpu?: number
  cpus?: number
  maxcpu?: number
  mem?: number
  maxmem?: number
  disk?: number
  maxdisk?: number
  uptime?: number
  template?: number
  diskread?: number
  diskwrite?: number
  netin?: number
  netout?: number
  disk_read_rate?: number
  disk_write_rate?: number
  net_in_rate?: number
  net_out_rate?: number
  guest_agent?: {
    qga_enabled: boolean
    qga_available: boolean
    qga_ip_address?: string | null
    qga_os_name?: string | null
    qga_disks?: PveFilesystemUsage[]
    filesystem_available?: boolean
    filesystem_source?: 'qga' | 'ssh' | 'winrm' | null
    filesystem_disks?: PveFilesystemUsage[]
    filesystem_error?: string | null
  }
  disk_rate_unavailable?: boolean
  net_rate_unavailable?: boolean
  [key: string]: unknown
}

export type PveGuestOs = 'linux' | 'windows' | 'unknown'

export interface PveGuestBinding {
  id: number | null
  connection_id: number
  guest_type: 'qemu'
  vmid: number
  ip_address: string | null
  os_system: PveGuestOs
  // true 表示 os_system=linux 是后端排除法兜底(未确证),打开绑定窗口会自动探测纠正
  os_assumed?: boolean
  // 持久化的精确 OS 名(QGA 不可用时的兜底,迁移 0038);QGA 实时 qga_os_name 优先
  os_name?: string | null
  ssh_port: number
  winrm_port: number
  username: string | null
  has_password: boolean
  has_ssh_key: boolean
  enabled: boolean
  configured: boolean
  last_test_status: string | null
  last_test_error: string | null
  last_test_at: string | null
  qga_enabled: boolean
  qga_available: boolean
  qga_ip_address: string | null
  qga_os_system?: 'linux' | 'windows' | null
  qga_os_name?: string | null
  qga_disks: PveFilesystemUsage[]
  winrm_available: boolean
}

export interface PveRrdPoint {
  time: number
  cpu?: number
  mem?: number
  maxmem?: number
  diskread?: number
  diskwrite?: number
  netin?: number
  netout?: number
  [key: string]: unknown
}

export interface PveSnapshot {
  name: string
  description?: string
  snaptime?: number
  vmstate?: number
  parent?: string
  [key: string]: unknown
}

export interface PveGuestCreate {
  vmid?: number
  name: string
  cores: number
  memory_mb: number
  disk_gb: number
  storage: string
  iso?: string
  /** 默认 true:QGA 是 IP/OS/磁盘采集自动化链路的开关 */
  agent?: boolean
}

/** 虚机引用的一个磁盘/光驱/挂载。kind 决定销毁虚机时数据会不会一起被清掉。 */
export interface PveGuestVolume {
  key: string
  volid: string
  size_bytes: number | null
  /** disk=虚机磁盘卷 · unused=残留卷 · cdrom=ISO 介质 · bind=LXC 宿主机绑定挂载 */
  kind: 'disk' | 'unused' | 'cdrom' | 'bind'
}

/** 删除前预览：会清掉哪些卷、会顺带清掉哪些平台侧引用。 */
export interface PveGuestVolumePreview {
  node: string
  status: string
  running: boolean
  template: boolean
  volumes: PveGuestVolume[]
  purged_local: {
    bindings: number
    containers: number
    business_links: number
  }
}

export interface PveGuestDeleteResult {
  success: boolean
  message: string
  task?: string | null
  node: string | null
  stopped_first: boolean
  volumes: PveGuestVolume[]
  purged_local: {
    bindings: number
    containers: number
    docker_status: number
    business_links: number
  }
}

/** PVE 异步任务(创建/克隆/销毁/快照)的执行状态。 */
export interface PveTaskStatus {
  /** running | stopped */
  status: 'running' | 'stopped'
  /** PVE 终态结论:OK 或错误文本;运行中为 null */
  exitstatus: string | null
  /** 任务是否成功(仅终态有意义) */
  ok: boolean
  /** 是否已到终态 */
  done: boolean
}

/** 创建虚机/容器的响应(任务提交回执)。 */
export interface PveGuestCreateResult {
  success: boolean
  vmid: number
  message: string
  task: string | null
  node: string
}

/** 克隆虚机/容器的响应(任务提交回执)。 */
export interface PveGuestCloneResult {
  success: boolean
  newid: number
  message: string
  task: string | null
  node: string
}

/** 节点存储(新建虚机的存储下拉数据源)。 */
export interface PveNodeStorage {
  storage: string
  type: string
  /** 逗号分隔的 content 列表,如 "images,iso" */
  content: string
  active: number
  total?: number
  used?: number
  avail?: number
}

/** 调整虚机规格的请求(未提供的字段不变;磁盘只允许扩容)。 */
export interface PveGuestConfigUpdate {
  cores?: number
  memory_mb?: number
  /** 磁盘目标总量(GB),必须大于当前容量 */
  disk_gb?: number
}

/** 调整规格的响应(带最新 config/status 供刷新)。 */
export interface PveGuestConfigUpdateResult {
  success: boolean
  message: string
  config: Record<string, unknown>
  disk_bytes: number | null
  status: Record<string, unknown>
}
